feat(table): migrate to core tokens

jawinn opened the pull request:
## Description

Use new 5.0 beta version of spectrum-css/table that includes the core tokens migration, section header, and additional variants. Modifies configuration to handle new CSS. Does not yet integrate all new features (except for `emphasized`, `quiet`, and `density`). 

Component changes:
- **headCell**: Adds a 'headCell' property / 'head-cell' attribute to the checkbox cell component. And modifies the spectrum-config to include the various styles depending on whether that attribute exists or not. Previously all styles for Table-cell, Table-headCell, and Table-checkboxCell were being applied to the sp-table-checkbox-cell, regardless of where it was located. Resulting in head cell styles appearing on checkbox body cells, and other incorrect styles.
- **emphasized**: Adds an emphasized attribute which applies emphasized or non-emphasized colors to selected rows.
- **density**: Adds the density attribute for compact and spacious.
- **quiet**: Adds support for the quiet attribute.
- **t-shirt size**: Adds support for large and extra large, and adds size section to docs examples.
- Sort icon moved the the left side

Fixes #3400 

## Related issue(s)

See [spectrum-css table PR](https://github.com/adobe/spectrum-css/pull/1842) for details.

## How has this been tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to see how your change affects other areas of the code, etc. -->

## Screenshots

![Screenshot 2023-08-17 at 4 27 58 PM](https://github.com/adobe/spectrum-web-components/assets/965114/fe780e50-17ab-4c4c-8df7-9ec9b7a5feab)

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [x] Breaking change (fix or feature that would cause existing functionality to change)
-   [ ] Chore (minor updates related to the tooling or maintenance of the repository, does not impact compiled assets)

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply.  If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

-   [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
-   [x] My code follows the code style of this project.
-   [ ] If my change required a change to the documentation, I have updated the documentation in this pull request.
-   [x] I have read the **[CONTRIBUTING](<(https://github.com/adobe/spectrum-web-components/blob/main/CONTRIBUTING.md)>)** document.
-   [ ] I have added tests to cover my changes.
-   [ ] All new and existing tests passed.
-   [ ] I have reviewed at the Accessibility Practices for this feature, see: [Aria Practices](https://www.w3.org/TR/wai-aria-practices/)

## Best practices

This repository uses conventional commit syntax for each commit message; note that the GitHub UI does not use this by default so be cautious when accepting suggested changes. Avoid the "Update branch" button on the pull request and opt instead for rebasing your branch against `main`.

jawinn commented:
@najikahalsema I'm seeing a few styling issues related to the _sp-table-checkbox-cell_:
![Screenshot 2023-07-19 at 3 41 37 PM](https://github.com/adobe/spectrum-web-components/assets/965114/1d3a3811-a0e7-4888-9a40-ddfa32e11d46)

It is missing the following, which causes extra spacing in the head checkbox cell:
`.spectrum-Table-checkboxCell.spectrum-Table-headCell { padding-block: 0; }`
An extra border is appearing at the top, which is strange. I'm not sure where that is coming from.

In Spectrum CSS, both a head cell and a regular cell can have the `.spectrum-Table-checkboxCell` class applied to them. Then, there are some style rules like the following:
```
.spectrum-Table-checkboxCell.spectrum-Table-cell { ... }
.spectrum-Table-checkboxCell.spectrum-Table-headCell { ... }
```

What would be needed here to make sure the right styles are coming through? Does the structure need to change at all, or does something need to be done to the spectrum-config.js? I see SWC has a separate component for the checkbox cell used in both the head and body.

najikahalsema commented:
> @najikahalsema I'm seeing a few styling issues related to the _sp-table-checkbox-cell_: ![Screenshot 2023-07-19 at 3 41 37 PM](https://user-images.githubusercontent.com/965114/254699611-1d3a3811-a0e7-4888-9a40-ddfa32e11d46.png)
> 
> It is missing the following, which causes extra spacing in the head checkbox cell: `.spectrum-Table-checkboxCell.spectrum-Table-headCell { padding-block: 0; }` An extra border is appearing at the top, which is strange. I'm not sure where that is coming from.
> 
> In Spectrum CSS, both a head cell and a regular cell can have the `.spectrum-Table-checkboxCell` class applied to them. Then, there are some style rules like the following:
> 
> ```
> .spectrum-Table-checkboxCell.spectrum-Table-cell { ... }
> .spectrum-Table-checkboxCell.spectrum-Table-headCell { ... }
> ```
>
> What would be needed here to make sure the right styles are coming through? Does the structure need to change at all, or does something need to be done to the spectrum-config.js? I see SWC has a separate component for the checkbox cell used in both the head and body.

Perhaps those classes are too specific? The checkbox on main is only getting styles from the checkboxCell css, rather than inheriting them from another parent component. I would try excluding the checkbox-related styling classes in locations other than table-checkboxCell CSS and see what happens. For the head checkbox cell, You could try adding the classes to it imperatively when it's first created (around line 301 in Table.ts). The `.checkbox` class will already exist because it's part of the render function of that component, so it'd be a matter of including `.spectrum-Table-headCell `as a class name.

I'd also take a look at the differences between the styling of the checkboxes in the head and body of the table. Are there a lot of similarities there? Where are the differences? Could they be combined into the already-existing `.checkbox` class? If so, perhaps it's worth adjusting the imported CSS to make that a reality.

jawinn commented:
@najikahalsema I made some changes to add a `headCell` boolean property, and adjusted all of the spectrum-config so that the checkbox cell styles includes Table-cell or Table-headCell styles based on whether the `head-cell` attribute exists or not. That seems to have cleared up most of those issues. Does this look alright? [6fb1c76](https://github.com/adobe/spectrum-web-components/pull/3441/commits/6fb1c767bf0b6a0342d8c754da863135f80b0460) (_note there is one slightly off `#box` style that will be changing as I'll be simplifying that with a custom property pass-through instead in the next beta version_).

najikahalsema commented:
Looks good to me! I wouldn't worry about the repeated `focused` and `focus-visible` styles for now as that's common for our other components as well.

jawinn commented:
Things are looking pretty good here now! I'll be posting for review. 
I just had one remaining question:

-----
Looking at achieving parity with production at the moment, I noticed something odd with the docs and how scrolling is used for non-virtualized. Looking at [production](https://opensource.adobe.com/spectrum-web-components/components/table/)—

At the start of the docs it says:
> To ensure that the table scrolls, make sure to add a style attribute to <sp-table> with your desired height. Otherwise, the table will automatically show all its items.

And then several of the non-virtualized tables set a height, creating a tiny bit of scroll but not in a fixed header sort of way shown in the virtualized examples. See the first "Selection" example. How this occurring is actually by the table visibly overflowing itself, with the scrolling actually happening on the wrapper that has `<div class="demo-example">` and `overflow: auto`, which is not much different behavior than just putting a wrapping div around it with a fixed height and `overflow: auto` (well, it's a little worse because the bottom padding gets lost).

Thoughts on options for updating:
  a. Remove the quoted part of the docs, along with the fixed heights on the non-virtualized+scroller examples. This isn't really currently showing a feature of table.
  b. Show a separate scrolling example under the first example with a wrapper and a fixed height and `overflow: auto`, and remove heights from the non-virtualized+scroller examples (seems not necessary but would be like what is demoed now)
  c. Add scroller support for non-virtualized and move "_The scroller property_" docs up to the top under a second example; if this is not a simple update or wouldn't be possible, it'd probably be better to wait since the integration of the latest scrollable wrapper changes in Spectrum CSS would change the implementation of this feature completely (and allow scrolling for both).

@najikahalsema @Westbrook

Rajdeepc commented on packages/table/src/table-head-cell.css:
```diff
@@ -15,4 +15,5 @@ governing permissions and limitations under the License.
 :host {
     display: block;
     flex: 1;
+    block-size: auto;
```
can't this come from css?

jawinn replied:
It's needed in this case because the display is changed to `flex`, which changes the behavior so it no longer works like a min-height like it does with `display: table-cell`.

Rajdeepc commented on packages/table/stories/table-virtualized.stories.ts:
```diff
@@ -200,6 +202,7 @@ virtualizedSingle.args = {
     return html`
         <sp-table
+            emphasized
             size="m"
```
if not size is mentioned what would be the default?

jawinn replied:
In CSS, all medium styles display as default.

jawinn commented on packages/table/src/Table.ts:
```diff
@@ -223,27 +230,33 @@ export class Table extends SizedMixin(SpectrumElement, {
             this.manageHeadCheckbox(allSelected);
         } else {
+            // Remove all checkbox cells.
             checkboxes.forEach((box) => {
```
What would you recommend? I just added a comment here and wasn't changing this code.

jawinn commented on packages/table/src/Table.ts:
```diff
@@ -298,27 +311,32 @@ export class Table extends SizedMixin(SpectrumElement, {
+            this.tableHeadCheckboxCell.headCell = true;
+            this.tableHeadCheckboxCell.emphasized = this.emphasized;
 
+            const allSelected = this.selected.length === this.tableRows.length;
```
I think this existing line is okay? It looks like `this.selected` is initialized as an empty array, and the `this.tableRows` is a getter that will return an empty array or the results of `querySelectorAll`, which is a NodeList (which has a `.length` property).

jawinn commented on packages/table/src/spectrum-config.js:
```diff
@@ -31,10 +31,14 @@ const config = {
             fileName: 'table',
             components: [
                 converter.classToHost(),
+                converter.classToAttribute('spectrum-Table--quiet'),
```
Sure. Adding density should be pretty straightforward with a new property, as all of the density changes in CSS are custom property value changes on the root element.

jawinn replied:
I've just added the `density` property and docs for it.

jianliao replied:
@jawinn Does this line indicate that `quiet` mode will be supported?

jawinn replied:
@jianliao I've added support for the `quiet` variant (including docs) in the latest commits.

najikahalsema commented:
> Things are looking pretty good here now! I'll be posting for review. I just had one remaining question:
> 
> Looking at achieving parity with production at the moment, I noticed something odd with the docs and how scrolling is used for non-virtualized. Looking at [production](https://opensource.adobe.com/spectrum-web-components/components/table/)—
> 
> At the start of the docs it says:
> 
> > To ensure that the table scrolls, make sure to add a style attribute to  with your desired height. Otherwise, the table will automatically show all its items.
> 
> And then several of the non-virtualized tables set a height, creating a tiny bit of scroll but not in a fixed header sort of way shown in the virtualized examples. See the first "Selection" example. How this occurring is actually by the table visibly overflowing itself, with the scrolling actually happening on the wrapper that has `<div class="demo-example">` and `overflow: auto`, which is not much different behavior than just putting a wrapping div around it with a fixed height and `overflow: auto` (well, it's a little worse because the bottom padding gets lost).
> 
> Thoughts on options for updating: a. Remove the quoted part of the docs, along with the fixed heights on the non-virtualized+scroller examples. This isn't really currently showing a feature of table. b. Show a separate scrolling example under the first example with a wrapper and a fixed height and `overflow: auto`, and remove heights from the non-virtualized+scroller examples (seems not necessary but would be like what is demoed now) c. Add scroller support for non-virtualized and move "_The scroller property_" docs up to the top under a second example; if this is not a simple update or wouldn't be possible, it'd probably be better to wait since the integration of the latest scrollable wrapper changes in Spectrum CSS would change the implementation of this feature completely (and allow scrolling for both).
> 
> @najikahalsema @Westbrook

If the feature is not really a feature, you can go ahead and remove it a la option A and see where that takes us.

jawinn commented:
> ...
> If the feature is not really a feature, you can go ahead and remove it a la option A and see where that takes us.

Sounds good. I've updated this.

I also updated the docs around the t-shirt sizing, with medium now being the default. I added a new section for Sizing based on how it's set up in some other components' docs, and have included support for large and extra-large.

Westbrook commented on packages/table/src/spectrum-table.css:
```diff
@@ -12,658 +12,561 @@ governing permissions and limitations under the License.
+        }
+    }
+}
+:host:not(.spectrum-Table-scroller) {
+    border-collapse: separate;
+    border-spacing: 0;
+}
+:host:not(.spectrum-Table-scroller) {
+    display: table;
+}
```
What do we need to do or document about these?

jawinn replied:
Nothing at the moment. Unless you think it's necessary to change the spectrum config to make these just `:host` for now. The `:not` is connected to the [scroller](https://pr-1842--spectrum-css.netlify.app/table#scrollablewithfixedcolumnheaders) that is not yet implemented on the SWC side (which would remove the need to use flex display properties and has an additional wrapper). I don't know if and how you want to implement that feature yet; this PR brings things to parity with the existing flex display w/ scroller.

Rajdeepc reviewed (commented):
Looks like you have addressed most of the review notes. Except for the merge conflicts we are good here on this.

najikahalsema commented on packages/table/README.md:
Could you remove the `emphasized` attribute from the `selects="single"` and `selects="multiple"` examples? If you do that, I don't think it's necessary to have a non-emphasized example in the docs site, since non-emphasized is the default. Correct me if I'm wrong and we want to... ahem, _emphasize_ the default to be `emphasized`, though. Also, I believe if you rebase to main, the Virtualised examples here should be fixed.

jawinn replied:
All of these comments should now be addressed in the latest commit.

najikahalsema commented on packages/table/README.md:
````diff
@@ -141,8 +138,7 @@ When `selects="multiple"`, the `<sp-table>` will manage a selection in the array
 
 ```html
````
I don't know why it's not letting me leave a comment on line 137, but anyways... while you're working on this PR, could you fix the first sentence in that line for me? 😅 

> When `selects="multiple"`, the `<sp-table>` manages selections via a presence toggle and adds them to the `selected` array.

najikahalsema commented on packages/table/stories/table-elements.stories.ts:
```diff
@@ -53,7 +53,7 @@ export default {
 export const elements = (): TemplateResult => {
     return html`
-        <sp-table size="m">
+        <sp-table emphasized>
```
again, I wonder if we want a majority of the examples to be emphasized, rather than the other way around, ie only specifying "emphasized" for a single story.

jawinn replied:
I'm not sure which one we should be showing in most default examples. One thought to add to the discussion;

The default for table previously had the selected row in blue. With the addition of emphasized and non-emphasized, the default selected row color will be grey. So in one way, the previous default looked more like emphasized (the checkbox did not have emphasized colors though).

najikahalsema replied:
That's interesting. I never realised that the checkbox and highlighted rows were different colours LOL 

I think I'd err on the side of matching our other component stories, which usually have a singular "emphasized" story and leave the rest as-is.

najikahalsema reviewed (changes requested):
So close! Just a few little things regarding `emphasized`. Feel free to disagree with me. 

Once you do that, rebase to main, update your CI hash, and I will approve. :)

Westbrook commented on packages/table/test/virtualized-table-selects.test.ts:
```diff
@@ -500,8 +505,8 @@ describe('Virtualized Table Selects', () => {
-        await oneEvent(el, 'rangeChanged');
-        await elementUpdated(el);
+        // await oneEvent(el, 'rangeChanged');
+        // await elementUpdated(el);
```
Comments.

Westbrook commented on packages/table/test/virtualized-table-selects.test.ts:
```diff
@@ -458,9 +460,13 @@ describe('Virtualized Table Selects', () => {
         const el = test.querySelector('sp-table') as Table;
+        const body = el.querySelector(
+            'sp-table-body'
+        ) as unknown as TableBody & {
+            [virtualizerRef]: Virtualizer;
+        };
 
-        await oneEvent(el, 'rangeChanged');
-        await elementUpdated(el);
+        await body[virtualizerRef].layoutComplete;
```
Should we make this:
```
await tableLayoutComplete(el);
```

And then have a helper at the top or in a different file like:

```
async function tableLayoutComplete(table: Table): Promise<void> {
    const body = el.querySelector(
            'sp-table-body'
        ) as unknown as TableBody & {
            [virtualizerRef]: Virtualizer;
        };
    await body[virtualizerRef].layoutComplete; 
}
```

Can accept this as follow up work if you'll link the issue you create here.

najikahalsema replied:
Nah, I can do it! Thank you!

Westbrook reviewed (commented):
Awesome to see this coming in for a landing! 🛬 

Two nits and I'm good to go!

Westbrook commented on packages/table/test/virtualized-table-selects.test.ts:
```diff
@@ -162,13 +148,12 @@ describe('Virtualized Table Selects', () => {
             `
         );
         const el = test.querySelector('sp-table') as Table;
+        const body = el.querySelector('sp-table-body') as unknown as TableBody;
```
```suggestion
        const body = el.querySelector('sp-table-body') as TableBody;
```

Westbrook commented on packages/table/test/virtualized-table-selects.test.ts:
```diff
@@ -240,19 +223,15 @@ describe('Virtualized Table Selects', () => {
             `
         );
         const el = test.querySelector('sp-table') as Table;
+        const body = el.querySelector('sp-table-body') as unknown as TableBody;
```
```suggestion
        const body = el.querySelector('sp-table-body') as TableBody;
```

Westbrook commented on packages/table/test/virtualized-table-selects.test.ts:
```diff
@@ -458,9 +421,9 @@ describe('Virtualized Table Selects', () => {
             `
         );
         const el = test.querySelector('sp-table') as Table;
+        const body = el.querySelector('sp-table-body') as unknown as TableBody;
```
```suggestion
        const body = el.querySelector('sp-table-body') as TableBody;
```

Westbrook reviewed (approved):
Thanks for driving this home @jawinn! Lots of great new features included herein. And, @najikahalsema the last minute testing polish is gonna serve us well in the future. :shipit:
